synapse/handlers/device.py

```diff
@@ -941,6 +944,9 @@ async def user_device_resync(
             await self.store.update_remote_device_list_cache(
                 user_id, devices, stream_id
             )
+        # mark the cache as valid, whether or not we actually processed any device
+        # list updates.
+        await self.store.mark_remote_user_device_cache_as_valid(user_id)
```

Comment:
the objective here is to move the `delete` outside the `if not ignore_devices` condition. However, I'm not *certain* it's safe to move it outside the `update_remote_device_list_cache` transaction. @erikjohnston I think this is your code - thoughts welcome.

Reply:
The only concern I have is if we mark the list as stale again between fetching the devices and checking against the cache? But if that is a problem I don't see how it isn't also a problem in the case where we do update the local device cache.